crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,771 @@
+    #[inline]
+    unsafe fn get(&self, index: u32) -> Entity {
+        // SAFETY: caller ensure we are init.
+        let head = unsafe { self.ptr().debug_checked_unwrap() };
+        // SAFETY: caller ensures we are in bounds (because `set` must be in bounds)
+        let target = unsafe { &*head.add(index as usize) };
+
+        target.get_entity()
+    }
+
+    /// Gets a slice of indices.
+    ///
+    /// # Safety
+    ///
+    /// [`Self::set`] must have been called on these indices before.
+    #[inline]
+    unsafe fn get_slice(&self, index: u32, ideal_len: u32, index_of_self: u32) -> &[Slot] {
+        let cap = Self::capacity_of_chunk(index_of_self);
+        let after_index_slice_len = cap - index;
+        let len = after_index_slice_len.min(ideal_len) as usize;
+
+        // SAFETY: caller ensure we are init.
+        let head = unsafe { self.ptr().debug_checked_unwrap() };
+
+        // SAFETY: The chunk was allocated via a `Vec` and the index is within the capacity.
+        unsafe { core::slice::from_raw_parts(head, len) }
+    }
+
+    /// Sets this entity at this index.
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    /// Index must be in bounds.
+    /// Access does not conflict with another [`Self::get`].
+    #[inline]
+    unsafe fn set(&self, index: u32, entity: Entity, index_of_self: u32) {
+        let head = self.ptr().unwrap_or_else(|| self.init(index_of_self));
+        // SAFETY: caller ensures it is in bounds and we are not fighting with other `set` calls or `get` calls.
+        // A race condition is therefore impossible.
+        let target = unsafe { &*head.add(index as usize) };
+
+        target.set_entity(entity);
+    }
+
+    /// Initializes the chunk to be valid, returning the pointer.
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    #[cold]
+    unsafe fn init(&self, index: u32) -> *mut Slot {
+        let cap = Self::capacity_of_chunk(index);
+        let mut buff = ManuallyDrop::new(Vec::new());
+        buff.reserve_exact(cap as usize);
+        let ptr = buff.as_mut_ptr();
+        self.first.store(ptr, Ordering::Relaxed);
+        ptr
+    }
+
+    /// Frees memory
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    unsafe fn dealloc(&self, index: u32) {
+        if let Some(to_drop) = self.ptr() {
+            let cap = Self::capacity_of_chunk(index) as usize;
+            // SAFETY: This was created in [`Self::init`] from a standard Vec.
+            unsafe {
+                Vec::from_raw_parts(to_drop, cap, cap);
+            }
+        }
+    }
+
+    /// Returns [`Self::first`] if it is valid.
+    #[inline]
+    fn ptr(&self) -> Option<*mut Slot> {
+        let ptr = self.first.load(Ordering::Relaxed);
+        (!ptr.is_null()).then_some(ptr)
+    }
+
+    fn new() -> Self {
+        Self {
+            first: AtomicPtr::new(core::ptr::null_mut()),
+        }
+    }
```

Comment:
Move `new` to be first. It sets the stage that we expect to init this separately.